### PR TITLE
Add support for OpenSSL 1.0.2

### PR DESCRIPTION
**Pull request by SueChaplain**
JDK 12 only

OpenSSL 1.0.2 is available on all platforms for Digest, CBC, GCM, and RSA.
On Linux and AIX you need to have it in your system path. For other
platforms the Adopt builds will bundle the 1.1.X library.

Digest re-enabled for 1.1.x

Closes: #188

Signed-off-by: Sue Chaplain <sue_chaplain@uk.ibm.com>

### CODE REVIEW

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/djdknativedigest.md [hunk 1]
```diff
@@ -42,9 +41,9 @@ This option enables or disables OpenSSL native cryptographic support for the Dig
 
 OpenSSL support is enabled by default for the CBC, Digest, GCM, and RSA algorithm. If you want to turn off the Digest algorithm, set this option to `false`.
 
-To turn off all the algorithms, see the [-Djdk.nativeCrypto](djdknativecrypto.md) system property command line option.-->
-
+To turn off all the algorithms, see the [-Djdk.nativeCrypto](djdknativecrypto.md) system property command line option.
 
+[End of content that applies only to Java 12](cr/java_close.png)
```
Comment (pshipton) [inline docs/djdknativedigest.md]: This tag isn't working. Missing ! ?

File: docs/introduction.md [hunk 2]
```diff
@@ -70,14 +70,18 @@ To improve the performance of applications that run in containers, try setting t
 OpenJDK uses the in-built Java cryptographic implementation by default. However, native cryptographic implementations
 typically provide better performance. OpenSSL is a native open source cryptographic toolkit for Transport Layer Security (TLS) and
 Secure Sockets Layer (SSL) protocols, which is well established and used with many enterprise applications. The OpenSSL V1.1.x implementation is
-currently supported for the <!--Digest,--> CBC, GCM, and RSA algorithms.
+currently supported for the ![Start of content that applies only to Java 12](cr/java12.png)Digest![End of content that applies only to Java 12](cr/java_close.png), CBC, GCM, and RSA algorithms.
 
-OpenSSL support is enabled by default for all these supported algorithms. If you want to limit support to specific algorithms, a number of
+![Start of content that applies only to Java 12](cr/java12.png)The OpenSSL V1.0.2 implementation is also supported for the Digest, CBC, GCM, and
+RSA algorithms. On Linux and AIX platforms, the OpenSSL 1.0.2 or 1.1.X library is expected to be found on the system path. (LIBPATH for Linux;
```
Comment (pshipton) [inline docs/introduction.md]: It doesn't need to be on the LIBPATH, etc. On Linux when you install the openssl package the ldconfig cache is updated to include the libcrypto library. I assume AIX has some similar mechanism.

File: docs/version0.13.md [hunk 3]
```diff
@@ -41,10 +43,21 @@ OpenJ9 release 0.13.0 supports OpenJDK 12, which is available from the AdoptOpen
 
 OpenJDK 12 with Eclipse OpenJ9 is not a long term support (LTS) release.
 
-Although it is possible to build an OpenJDK 8, or other versions, with OpenJ9 0.13.0, testing at the project is not complete and therefore support is not available.
+The latest builds of OpenJDK with OpenJ9 for Java 8 and 11 at the AdoptOpenJDK community are for Eclipse OpenJ9 release 0.12.0. Features mentioned in these release notes are not available in these builds. Although it is possible to build an OpenJDK 8 or OpenJDK 11 with OpenJ9 0.13.0, testing at the project is not complete and therefore support for any of these features is not available.
```
Comment (pshipton) [inline docs/version0.13.md]: "may" be possible to build. Although it works today, a future change could break it at any time.

File: docs/version0.13.md [hunk 4]
```diff
@@ -41,10 +43,21 @@ OpenJ9 release 0.13.0 supports OpenJDK 12, which is available from the AdoptOpen
...
 To learn more about support for OpenJ9 releases, including OpenJDK levels and platform support, see [Supported environments](openj9_support.md)
 
+### Support for OpenSSL 1.0.2
+
+![Start of content that applies only to Java 12](cr/java12.png)
+
+OpenSSL cryptographic support is extended to include OpenSSL 1.0.2 for the Digest, CBC, GCM, and RSA algorithms. Support is enabled by default
+and the OpenSSL libraries are expected to be available on the system library path. For more information about cryptographic acceleration with OpenSSL, see [Cryptographic operations](introduction.md#cryptographic-operations).
+
+In addition, support for OpenSSL 1.1.X is re-enabled in this release following the resolution of issue [#4530](https://github.com/eclipse/openj9/issues/4530).
```
Comment (pshipton) [inline docs/version0.13.md]: Support for NativeDigest is re-enabled, not OpenSSL 1.1.X

File: docs/djdknativedigest.md (hunk 1 above)
Comment (SueChaplain) [reply docs/djdknativedigest.md]: It was missing the "!". Now fixed.

File: docs/introduction.md (hunk 2 above)
Comment (SueChaplain) [reply docs/introduction.md]: The input I had from Nasser was: 

>The crypto library is not bundled by default. User can set the PATH or LIBPATH or LD_LIBRARY_PATH to crypto library if the library is not already in the system path.

This implies that there might be some situations where you do need to update those env vars. I've modified the text, but if you think I should remove these mentions altogether, I can do.

File: docs/version0.13.md (hunk 3 above)
Comment (SueChaplain) [reply docs/version0.13.md]: Changed to "might be"

File: docs/version0.13.md (hunk 4 above)
Comment (SueChaplain) [reply docs/version0.13.md]: Yep ... fixed. Thanks for your review. I gave up waiting for it to stage before I left yesterday so had not been through it.

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/introduction.md (hunk 2 above)
Comment (pshipton) [reply docs/introduction.md]: While you can modify the LD_LIBRARY_PATH to include it, it really should be installed via the OS package manager, which uses the ldconfig. Please recommend using the OS package manager first. Not sure we should be recommending manually adding it, but note Linux is LD_LIBRARY_PATH and AIX is LIBPATH.

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/introduction.md (hunk 2 above)
Comment (SueChaplain) [reply docs/introduction.md]: I've changed the wording, removing the setting of any env vars if we are not sure that should be recommended. I guess we have to wait to see where things get to on AIX before merging.

Comment (pshipton) [reply docs/introduction.md]: AIX is "working", although the changes to make it work haven't been merged yet.

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/introduction.md (hunk 2 above)
Comment (pshipton) [reply docs/introduction.md]: The AIX changes are merged now.

Comment (SueChaplain) [reply docs/introduction.md]: Do you think this is ready to merge now @pshipton? If so, please go ahead. Latest work staged here: https://eclipse.github.io/openj9-docs-staging/221/introduction/#cryptographic-operations

File: docs/version0.13.md [hunk 5]
```diff
@@ -41,15 +43,26 @@ OpenJ9 release 0.13.0 supports OpenJDK 12, which is available from the AdoptOpen
...
+
+OpenSSL cryptographic support is extended to include OpenSSL 1.0.2 for the Digest, CBC, GCM, and RSA algorithms. Support is enabled by default. On Linux and AIX platforms, the OpenSSL libraries are expected to be available on the system path. For more information about cryptographic acceleration with OpenSSL, see [Cryptographic operations](introduction.md#cryptographic-operations).
+
+In addition, support for the OpenSSL 1.1.X Digest algorithm is re-enabled in this release following the resolution of issue [#4530](https://github.com/eclipse/openj9/issues/4530).
+
+![End of content that applies only to Java 12](cr/java_close.png)
 
 ### New Java process status tool
 
 ![Start of content that applies only to Java 12](cr/java12.png)
```
Comment (pshipton) [inline docs/version0.13.md]: This jdk12 content is never closd.

File: docs/introduction.md (hunk 2 above)
Comment (pshipton) [reply docs/introduction.md]: One other issue to resolve.

**Comment by SueChaplain** [comment]
@pshipton  - I believe this is ready too.

**Comment by SueChaplain** [comment]
@pshipton I added a stronger warning into the OpenSSL section about the library not being bundled for JDK12. I'm still considering where we would put information about "pre-reqs". I'm thinking about a "Known issues and limitations topic" .... any thoughts?

**Comment by pshipton** [comment]
I'm thinking a new section underneath "New to Open?" called `Installation Prerequisites` where hopefully everybody will see it. Not sure anybody would look at `Known issues and limitations` unless they have a problem, and even then maybe not. For OpenSSL the problem is performance which is more subtle.

**Comment by SueChaplain** [comment]
@pshipton - is there anything left to do here please? The pre-reqs are now in that AdoptOpenJDK builds topic. If you think we are done, please merge.

File: docs/version0.13.md [hunk 6]
```diff
@@ -41,15 +43,29 @@ OpenJ9 release 0.13.0 supports OpenJDK 12, which is available from the AdoptOpen
...
-To learn more about support for OpenJ9 releases, including OpenJDK levels and platform support, see [Supported environments](openj9_support.md)
+To learn more about support for OpenJ9 releases, including OpenJDK levels and platform support, see [Supported environments](openj9_support.md).
+
+### Support for OpenSSL 1.0.2
+
+![Start of content that applies only to Java 12](cr/java12.png)
+
+OpenSSL cryptographic support is extended to include OpenSSL 1.0.2 for the Digest, CBC, GCM, and RSA algorithms. Support is enabled by default. On Linux and AIX platforms, the OpenSSL libraries are expected to be available on the system path. For more information about cryptographic acceleration with OpenSSL, see [Cryptographic operations](introduction.md#cryptographic-operations).
+
+In addition, support for the OpenSSL 1.1.X Digest algorithm is re-enabled in this release following the resolution of issue [#4530](https://github.com/eclipse/openj9/issues/4530).
```
Comment (pshipton) [inline docs/version0.13.md]: This should say "support for the OpenSSL Digest algorithm" and not "support for the OpenSSL **1.1.X** Digest algorithm"

Comment (SueChaplain) [reply docs/version0.13.md]: Fixed

**Pull request merged**